### PR TITLE
Add airdrop snapshot command

### PR DESCRIPTION
**Pull request by sunnya97**
Command that generates a quadratic fairdrop snapshot.json based on a provided cosmos-sdk v0.36 genesis export

Let's use this intermediate snapshot.json as the input into a future command the generates the genesis json (based off desired claim properties)

### CODE REVIEW

File: cmd/osmosisd/cmd/airdrop.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+			serverCtx := server.GetServerContextFromCmd(cmd)
+			config := serverCtx.Config
+
+			config.SetRoot(clientCtx.HomeDir)
+
+			denom := args[0]
+			genesisFile := args[1]
+			snapshotOutput := args[2]
+
+			osmoSupplyStr, err := cmd.Flags().GetString(flagOsmoSupply)
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:86]:
```suggestion
			// Parse CLI input for osmo supply
			osmoSupplyStr, err := cmd.Flags().GetString(flagOsmoSupply)
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+			if err != nil {
+				return fmt.Errorf("failed to get osmo total supply: %w", err)
+			}
+			osmoSupply, ok := sdk.NewIntFromString(osmoSupplyStr)
+			if !ok {
+				return fmt.Errorf("failed to parse osmo supply: %s", osmoSupplyStr)
+			}
+
+			genesisJson, err := os.Open(genesisFile)
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:95]:
```suggestion
                        // Read genesis file
			genesisJson, err := os.Open(genesisFile)
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+			var genStateV036 GenesisStateV036
+
+			setCosmosBech32Prefixes()
+			err = aminoCodec.UnmarshalJSON(byteValue, &genStateV036)
+			if err != nil {
+				return err
+			}
+
+			snapshot := make(map[string]SnapshotFields)
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:111]:
```suggestion
			// Produce the map of address to total atom balance, both staked and unstaked
			snapshot := make(map[string]SnapshotFields)
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+					unbondingAtoms = unbondingAtoms.Add(entry.Balance)
+				}
+
+				acc.AtomBalance = acc.AtomBalance.Add(unbondingAtoms)
+				acc.AtomUnstakedBalance = acc.AtomUnstakedBalance.Add(unbondingAtoms)
+
+				snapshot[address] = acc
+			}
+
+			validators := make(map[string]v036staking.Validator)
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:149]:
```suggestion
			// Make a map from validator operator address to the v036 validator type
			validators := make(map[string]v036staking.Validator)
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+// AppStateV036 is app state structure for app state
+type AppStateV036 struct {
+	Accounts []v036genaccounts.GenesisAccount `json:"accounts"`
+	Staking  v036staking.GenesisState         `json:"staking"`
+}
+
+// SnapshotFields provide fields of snapshot per account
+type SnapshotFields struct {
+	AtomAddress           string  `json:"atom_address"`
+	AtomBalance           sdk.Int `json:"atom_balance"`
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:36]:
```suggestion
        // Atom Balance = AtomStakedBalance + AtomUnstakedBalance
	AtomBalance           sdk.Int `json:"atom_balance"`
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	AtomStakedBalance     sdk.Int `json:"atom_staked_balance"`
+	AtomUnstakedBalance   sdk.Int `json:"atom_unstaked_balance"`
+	AtomStakedPercent     sdk.Dec `json:"atom_staked_percent"`
+	AtomOwnershipPercent  sdk.Dec `json:"atom_ownership_percent"`
+	OsmoNormalizedBalance sdk.Int `json:"osmo_balance_normalized"`
+	OsmoBalance           sdk.Int `json:"osmo_balance"`
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:42]:
```suggestion
        // OsmoBalance = sqrt( AtomBalance + OSMO_BONUS * AtomStakedBalance)
	OsmoBalance           sdk.Int `json:"osmo_balance"`
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	OsmoBalanceBonus      sdk.Int `json:"osmo_balance_bonus"`
+	OsmoBalanceBase       sdk.Int `json:"osmo_balance_base"`
+	OsmoPercent           sdk.Dec `json:"osmo_ownership_percent"`
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:45]:
```suggestion
        // OsmoBalanceBonus = OsmoBalanceBase * (1.5 * atom staked percent)
	OsmoBalanceBonus      sdk.Int `json:"osmo_balance_bonus"`
	// OsmoBalanceBase = sqrt(atom balance)
	OsmoBalanceBase       sdk.Int `json:"osmo_balance_base"`
	// OsmoPercent = OsmoNormalizedBalance / TotalOsmoSupply
	OsmoPercent           sdk.Dec `json:"osmo_ownership_percent"`
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+				if err != nil {
+					panic(fmt.Sprintf("failed to root atom balance: %s", err))
+				}
+				acc.OsmoBalanceBase = baseOsmo.RoundInt()
+
+				bonusOsmo := baseOsmo.Mul(onePointFive).Mul(stakedPercent)
+				acc.OsmoBalanceBonus = bonusOsmo.RoundInt()
+
+				allOsmo := baseOsmo.Add(bonusOsmo)
+				acc.OsmoBalance = allOsmo.RoundInt()
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:203]:
```suggestion
                                // OsmoBalance = sqrt( all atoms) * (1 + 1.5) * (staked atom percent) = 
				acc.OsmoBalance = allOsmo.RoundInt()
```

File: cmd/osmosisd/cmd/airdrop.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	Staking  v036staking.GenesisState         `json:"staking"`
+}
+
+// SnapshotFields provide fields of snapshot per account
+type SnapshotFields struct {
+	AtomAddress           string  `json:"atom_address"`
+	AtomBalance           sdk.Int `json:"atom_balance"`
+	AtomStakedBalance     sdk.Int `json:"atom_staked_balance"`
+	AtomUnstakedBalance   sdk.Int `json:"atom_unstaked_balance"`
+	AtomStakedPercent     sdk.Dec `json:"atom_staked_percent"`
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:39]:
```suggestion
        // AtomStakedPercent = AtomStakedBalance / AtomBalance
	AtomStakedPercent     sdk.Dec `json:"atom_staked_percent"`
```

**Review by ValarDragon — COMMENTED** [review]
LGTM aside from comment needs

**Comment by sunnya97** [comment]
@ValarDragon thanks.  Agreed, we should be adding more comments throughout the code.

**Comment by zmanian** [comment]
Thanks for sharing this!

**Pull request merged**